include/internal/quic_cc.h

```diff
@@ -14,146 +14,188 @@
+    OSSL_TIME   largest_acked_time;
+} OSSL_CC_ECN_INFO;
+
+/* Parameter (read-write): Maximum datagram payload length in bytes. */
+#define OSSL_CC_OPTION_MAX_DGRAM_PAYLOAD_LEN        1
+
+/* Diagnostic (read-only): current congestion window size in bytes. */
+#define OSSL_CC_OPTION_CUR_CWND_SIZE                2
+
+/* Diagnostic (read-only): minimum congestion window size in bytes. */
+#define OSSL_CC_OPTION_MIN_CWND_SIZE                3
+
+/* Diagnostic (read-only): current net bytes in flight. */
+#define OSSL_CC_OPTION_CUR_BYTES_IN_FLIGHT          4
+
+/* Diagnostic (read-only): method-specific state value. */
+#define OSSL_CC_OPTION_CUR_STATE                    5
+
+/*
+ * Congestion control abstract interface.
+ *
+ * This interface is broadly based on the design described in RFC 9002. However,
+ * the demarcation between the ACKM and the congestion controller does not
+ * exactly match that delineated in the RFC 9002 psuedocode. Where aspects of
+ * the demarcation involve the congestion controller accessing internal state of
+ * the ACKM, the interface has been revised where possible to provide the
+ * information needed by the congestion controller and avoid needing to give the
+ * congestion controller access to the ACKM's internal data structures.
+ *
+ * Particular changes include:
+ *
+ *   - In our implementation, it is the responsibility of the ACKM to determine
+ *     if a loss event constitutes persistent congestion.
+ *
+ *   - In our implementation, it is the responsibility of the ACKM to determine
+ *     if the ECN-CE counter has increased. The congestion controller is simply
+ *     informed when an ECN-CE event occurs.
+ *
+ * All of these changes are intended to avoid having a congestion controller
+ * have to access ACKM internal state.
+ *
+ */
+#define OSSL_CC_LOST_FLAG_PERSISTENT_CONGESTION     (1U << 0)
 
+typedef struct ossl_cc_method_st {
     /*
-     * Reset the congestion control state.
-     * |flags| to support different level of reset (partial/full).
+     * Instantiation.
      */
-    void (*reset)(OSSL_CC_DATA *ccdata, int flags);
+    OSSL_CC_DATA *(*new)(OSSL_TIME (*now_cb)(void *arg),
+                         void *now_cb_arg);
+
+    void (*free)(OSSL_CC_DATA *ccdata);
 
     /*
-     * Set number of packets exempted from CC - used for probing
-     * |numpackets| is a small value (2).
-     * Returns 0 on error, 1 otherwise.
+     * Reset of state.
      */
-    int (*set_exemption)(OSSL_CC_DATA *ccdata, int numpackets);
+    void (*reset)(OSSL_CC_DATA *ccdata);
 
     /*
-     * Get current number of packets exempted from CC.
-     * Returns negative value on error, the number otherwise.
+     * Escape hatch for option configuration.
+     *
+     * option_id: One of OSSL_CC_OPTION_*.
+     *
+     * value: The option value to set.
+     *
+     * Returns 1 on success and 0 on failure.
      */
-    int (*get_exemption)(OSSL_CC_DATA *ccdata);
+    int (*set_option_uint)(OSSL_CC_DATA *ccdata,
```

Comment:
Do we need to reinvent a simplified version of the OSSL_PARAM API for this?

By using the OSSL_PARAM APIs, it is possible to support a provided congestion controller which was the longer term plan I thought.

Efficiency concerns can be circumvented by guaranteeing that the params don't move in memory and caching the located pointer.  I.e. during initialisation:
```c
my_option_pointer = OSSL_PARAM_locate("my option");
```
and this can later be freely used _without_ doing the locate again.

This would likely imply functions returning gettable and settable lists of parameters that take the CC as an argument.  This approach dissociates the parameters from small integer indices which is a good thing for future maintainability.

Reply:
This isn';t a public API yet, so I don't think this is much of a concern at this stage.

I've considered these kinds of solutions to OSSL_PARAM's issues, but I don't see a solution as being in scope for this PR.

I think it would be quite confusing to start using the existing OSSL_PARAM type in a completely new way but only in some code. It makes the obvious way to use OSSL_PARAM up until now not work, so reusing the OSSL_PARAM type doesn't seem to carry much advantage.

Reply:
I do see this as a concern.  We build a parameter passing mechanism and you are inventing a new one.  Reuse don't duplicate.

I don't see it as confusing so long as the parameter documentation clearly states that the address of this parameter is fixed for the lifetime of "X".  Since this is internal and performance is not a goal of MVP, refetch the param each time -- we can adjust that later.

Reply:
I think this should be brought up to OTC as I agree with @paulidale.